core/src/main/scala/scala/pickling/Macros.scala

```diff
@@ -497,7 +496,7 @@ trait UnpicklerMacros extends Macro with UnpickleMacros with FastTypeTagMacros {
       // in the worst case we can do the same as the interpreted runtime does - just go for allocateInstance
 
       // pending fields are fields that are restored after instantiation (e.g., through field assignments)
-      val pendingFields = if (!canCallCtor) cir.fields else cir.fields.filter(fir =>
+      var pendingFields = if (!canCallCtor) cir.fields else cir.fields.filter(fir =>
```

Comment:
@jsuereth I share your sentiment. :-) Cleaning up this logic would be nice. I submitted the PR in the present form mainly because I ran out of time, but still wanted to share my bug fix with people.